### PR TITLE
Add public interfaces for Rules evaluation

### PR DESCRIPTION
**Pull request by dagnir**
## Motivation and Context
This ports the public interface of the rules engine from the internal package.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/EndpointRuleset.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package software.amazon.awssdk.core.rules;
+
+import java.io.InputStream;
+
+/**
+ * The set of rules that are used to compute the endpoint to use for a request.
+ */
+public final class EndpointRuleset {
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/EndpointRuleset.java:23]: SdkProtected/InternalApi? (Same for all classes here).

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/EndpointRuleset.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    /**
+     * Create an instance of this object using from its JSON representation.
+     */
+    public static String fromJson(InputStream is) {
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/EndpointRuleset.java:27]: Why String?

Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/EndpointRuleset.java:27]: Errr, that's a mistake!

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/EndpointRuleset.java (hunk 1 above)
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/EndpointRuleset.java:23]: +1

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/Value.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    public static Int integer(int value) {
+        return new Int(value);
+    }
+
+    public static Bool bool(boolean value) {
+        return new Bool(value);
+    }
+
+    public static Record record(Map<Identifier, Value> value) {
+        return new Record(value);
```
Comment (millems) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/rules/Value.java:80]: nit: naming we usually do is fromBoolean, fromInteger, etc. (e.g. with Document).

**Review by millems — APPROVED** [review]

**Pull request merged**